flink/v1.15/flink/src/main/java/org/apache/iceberg/flink/sink/IcebergSink.java

```diff
@@ -0,0 +1,672 @@
+ *               |                 \                                                |
+ *               |                  \                                               |
+ *               |                   \                                              |
+ * +-------+     | +-------------+    \  +---------------+        +---------------+ |
+ * | Map 2 | ==> | | writer 2 | =| ---  >| committer 2   |  --->  | post commit   | |
+ * +-------+     | +-------------+       +---------------+        +---------------+ |
+ *               |                                                                  |
+ *               +------------------------------------------------------------------+
+ * }</pre>
+ */
+public class IcebergSink
+    implements StatefulSink<RowData, StreamWriterState>,
+        WithPreWriteTopology<RowData>,
+        WithPreCommitTopology<RowData, FilesCommittable>,
+        WithPostCommitTopology<RowData, FilesCommittable> {
+  private static final Logger LOG = LoggerFactory.getLogger(IcebergSink.class);
+
+  private final TableLoader tableLoader;
+  private final Table table;
+  private final boolean overwrite;
+  private final boolean upsertMode;
+  private final DistributionMode distributionMode;
+  private final List<String> equalityFieldColumns;
+  private final List<Integer> equalityFieldIds;
+  private final FileFormat dataFileFormat;
+  private final int workerPoolSize;
+  private final long targetDataFileSize;
+  private final String uidPrefix;
+  private final Map<String, String> snapshotProperties;
+  private final RowType flinkRowType;
+  private Committer<FilesCommittable> committer;
+
+  public IcebergSink(
+      TableLoader tableLoader,
+      @Nullable Table newTable,
+      TableSchema tableSchema,
+      List<String> equalityFieldColumns,
+      @Nullable String uidPrefix,
+      ReadableConfig readableConfig,
+      Map<String, String> writeOptions,
+      Map<String, String> snapshotProperties) {
+    this.tableLoader = tableLoader;
+    this.equalityFieldColumns = equalityFieldColumns;
+    this.uidPrefix = uidPrefix == null ? "" : uidPrefix;
+    this.snapshotProperties = snapshotProperties;
+
+    Preconditions.checkNotNull(tableLoader, "Table loader shouldn't be null");
+
+    if (newTable == null) {
+      tableLoader.open();
+      try (TableLoader loader = tableLoader) {
+        this.table = loader.loadTable();
+      } catch (IOException e) {
+        throw new UncheckedIOException(
+            "Failed to load iceberg table from table loader: " + tableLoader, e);
+      }
+    } else {
+      this.table = newTable;
+    }
+
+    FlinkWriteConf flinkWriteConf = new FlinkWriteConf(table, writeOptions, readableConfig);
+
+    this.equalityFieldIds = checkAndGetEqualityFieldIds(table, equalityFieldColumns);
+
+    this.flinkRowType = toFlinkRowType(table.schema(), tableSchema);
+
+    // Validate the equality fields and partition fields if we enable the upsert mode.
```

Comment:
nit: can we move the upsert validation to a separate method. might looks a little cleaner that way.

Reply:
done